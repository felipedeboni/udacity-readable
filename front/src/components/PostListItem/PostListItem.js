import React from 'react';
import { Link } from 'react-router-dom';
import PropTypes from 'prop-types';
import Moment from 'react-moment';
import FontAwesomeIcon from '@fortawesome/react-fontawesome';
import { faEdit, faTimes } from '@fortawesome/fontawesome-free-solid';
import { PostScoreContainer } from 'containers';
import react from 'react.svg';
import redux from './redux.svg';
import other from './other.svg';
import './PostListItem.css';

const PostListItem = ({
  post: { id, title, author, voteScore, commentCount, timestamp, category },
  removeHandler
}) => {
  const link = `/${category}/${id}`;
  let categoryLogo;
  let logoModifier = category;

  switch (category) {
    case 'react':
      categoryLogo = react;
      break;
    case 'redux':
      categoryLogo = redux;
      break;
    default:
      logoModifier = 'default';
      categoryLogo = other;
  }

  return (
    <li className="list__item">
      <div className="post__image">
        <Link to={link} className={`post__icon post__icon--${logoModifier}`}>
          <img src={categoryLogo} alt={category} />
        </Link>
      </div>
      <div className="list__item__data">
        <h2 className="post__title">
          <Link to={link}>{title}</Link>
        </h2>
        <div className="list__item__meta">
          <Moment
            className="post__date"
            date={timestamp}
            format="MMMM Do, YYYY"
          />
          <span className="separator"> • </span>
          <span>{author}</span>
          <span className="separator"> • </span>
          <span>
            {commentCount === 0 && 'No comments'}
            {commentCount > 0 && `${commentCount} comment`}
            {commentCount > 1 && 's'}
          </span>
        </div>
        <PostScoreContainer id={id} />
      </div>
      <div className="list__item__actions">
        <Link to={`${link}/edit`} className="list__item__actions__action">
          <FontAwesomeIcon icon={faEdit} />
        </Link>

        <a
          className="list__item__actions__action text-danger pointer"
          onClick={() => removeHandler(id)}
        >
          <FontAwesomeIcon icon={faTimes} />
        </a>
      </div>
    </li>
  );
};

PostListItem.propTypes = {
  post: PropTypes.shape({
    id: PropTypes.string.isRequired,
    title: PropTypes.string.isRequired,
    author: PropTypes.string.isRequired,
    voteScore: PropTypes.number.isRequired,
    commentCount: PropTypes.number.isRequired,
    timestamp: PropTypes.number.isRequired,
    category: PropTypes.string.isRequired
  }).isRequired,
  removeHandler: PropTypes.func.isRequired
};

export default PostListItem;
