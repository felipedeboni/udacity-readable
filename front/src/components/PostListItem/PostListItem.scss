.post__image {
  width: 30px;
  height: 30px;
  margin: 0 10px 0 0;
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: flex-start;
}

.post__icon {
  position: relative;
  width: 29px;
  height: 29px;
  border-radius: 50%;
  border: 2px solid #DDD;
  display: flex;
  align-items: center;
  justify-content: center;

  &--redux img {
    margin-top: -2px;
    width: 80%;
  }

  &--react img {
    margin-top: 1px;
    width: 100%;
  }

  &--default img {
    width: 50%;
  }
}

.post__title {
  font-size: 20px;
  font-weight: 700;
  margin: 0;
}

.post__title a {
  color: #54595F;

  &:hover {
    text-decoration: none;
  }
}

@media (min-width: 767px) {
  .post__image {
    width: 46px;
    height: 46px;
    margin: 0 15px 0 -2px;
  }

  .post__icon {
    width: 45px;
    height: 45px;
  }

  .post__title {
    font-size: 24px;
  }
}
