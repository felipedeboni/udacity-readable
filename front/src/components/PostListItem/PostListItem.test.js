import React from 'react';
import { shallow } from 'enzyme';
import PostListItem from './PostListItem';

const [postMock] = global.mock.posts;

describe('PostListItem', () => {
  it('renders without crashing', () => {
    const test = <PostListItem post={postMock} removeHandler={jest.fn} />;

    expect(shallow(test)).toMatchSnapshot();
  });

  it('renders the default category logo', () => {
    const post = {
      ...postMock,
      category: 'udacity'
    };
    const test = <PostListItem post={post} removeHandler={jest.fn} />;

    expect(shallow(test)).toMatchSnapshot();
  });

  it('renders the redux category logo', () => {
    const post = {
      ...postMock,
      category: 'redux'
    };
    const test = <PostListItem post={post} removeHandler={jest.fn} />;

    expect(shallow(test)).toMatchSnapshot();
  });

  it('renders the right string when there are no comments', () => {
    const post = {
      ...postMock,
      commentCount: 0
    };
    const wrapper = shallow(
      <PostListItem post={post} removeHandler={jest.fn} />
    );
    expect(wrapper.text()).toContain('No comments');
  });

  it('renders the right string when there are one comment', () => {
    const post = {
      ...postMock,
      commentCount: 1
    };
    const wrapper = shallow(
      <PostListItem post={post} removeHandler={jest.fn} />
    );
    expect(wrapper.text()).toContain('1 comment');
  });
});
