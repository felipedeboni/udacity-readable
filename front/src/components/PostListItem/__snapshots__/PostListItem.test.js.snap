// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostListItem renders the default category logo 1`] = `
<li
  className="list__item"
>
  <div
    className="post__image"
  >
    <Link
      className="post__icon post__icon--default"
      replace={false}
      to="/udacity/8xf0y6ziyjabvozdd253nd"
    >
      <img
        alt="udacity"
        src="other.svg"
      />
    </Link>
  </div>
  <div
    className="list__item__data"
  >
    <h2
      className="post__title"
    >
      <Link
        replace={false}
        to="/udacity/8xf0y6ziyjabvozdd253nd"
      >
        Udacity is the best place to learn React
      </Link>
    </h2>
    <div
      className="list__item__meta"
    >
      <t
        ago={false}
        calendar={false}
        className="post__date"
        date={1467166872634}
        element="time"
        filter={[Function]}
        format="MMMM Do, YYYY"
        fromNow={false}
        interval={60000}
        onChange={[Function]}
        toNow={false}
        unix={false}
        utc={false}
      />
      <span
        className="separator"
      >
         • 
      </span>
      <span>
        thingtwo
      </span>
      <span
        className="separator"
      >
         • 
      </span>
      <span>
        2 comment
        s
      </span>
    </div>
    <Connect(PostScoreContainer)
      id="8xf0y6ziyjabvozdd253nd"
    />
  </div>
  <div
    className="list__item__actions"
  >
    <Link
      className="list__item__actions__action"
      replace={false}
      to="/udacity/8xf0y6ziyjabvozdd253nd/edit"
    >
      <FontAwesomeIcon$1
        border={false}
        className=""
        fixedWidth={false}
        flip={null}
        icon={
          Object {
            "icon": Array [
              576,
              512,
              Array [],
              "f044",
              "M402.6 83.2l90.2 90.2c3.8 3.8 3.8 10 0 13.8L274.4 405.6l-92.8 10.3c-12.4 1.4-22.9-9.1-21.5-21.5l10.3-92.8L388.8 83.2c3.8-3.8 10-3.8 13.8 0zm162-22.9l-48.8-48.8c-15.2-15.2-39.9-15.2-55.2 0l-35.4 35.4c-3.8 3.8-3.8 10 0 13.8l90.2 90.2c3.8 3.8 10 3.8 13.8 0l35.4-35.4c15.2-15.3 15.2-40 0-55.2zM384 346.2V448H64V128h229.8c3.2 0 6.2-1.3 8.5-3.5l40-40c7.6-7.6 2.2-20.5-8.5-20.5H48C21.5 64 0 85.5 0 112v352c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48V306.2c0-10.7-12.9-16-20.5-8.5l-40 40c-2.2 2.3-3.5 5.3-3.5 8.5z",
            ],
            "iconName": "edit",
            "prefix": "fas",
          }
        }
        listItem={false}
        mask={null}
        name=""
        pull={null}
        pulse={false}
        rotation={null}
        size={null}
        spin={false}
        symbol={false}
        transform={null}
      />
    </Link>
    <a
      className="list__item__actions__action text-danger pointer"
      onClick={[Function]}
    >
      <FontAwesomeIcon$1
        border={false}
        className=""
        fixedWidth={false}
        flip={null}
        icon={
          Object {
            "icon": Array [
              384,
              512,
              Array [],
              "f00d",
              "M323.1 441l53.9-53.9c9.4-9.4 9.4-24.5 0-33.9L279.8 256l97.2-97.2c9.4-9.4 9.4-24.5 0-33.9L323.1 71c-9.4-9.4-24.5-9.4-33.9 0L192 168.2 94.8 71c-9.4-9.4-24.5-9.4-33.9 0L7 124.9c-9.4 9.4-9.4 24.5 0 33.9l97.2 97.2L7 353.2c-9.4 9.4-9.4 24.5 0 33.9L60.9 441c9.4 9.4 24.5 9.4 33.9 0l97.2-97.2 97.2 97.2c9.3 9.3 24.5 9.3 33.9 0z",
            ],
            "iconName": "times",
            "prefix": "fas",
          }
        }
        listItem={false}
        mask={null}
        name=""
        pull={null}
        pulse={false}
        rotation={null}
        size={null}
        spin={false}
        symbol={false}
        transform={null}
      />
    </a>
  </div>
</li>
`;

exports[`PostListItem renders the redux category logo 1`] = `
<li
  className="list__item"
>
  <div
    className="post__image"
  >
    <Link
      className="post__icon post__icon--redux"
      replace={false}
      to="/redux/8xf0y6ziyjabvozdd253nd"
    >
      <img
        alt="redux"
        src="redux.svg"
      />
    </Link>
  </div>
  <div
    className="list__item__data"
  >
    <h2
      className="post__title"
    >
      <Link
        replace={false}
        to="/redux/8xf0y6ziyjabvozdd253nd"
      >
        Udacity is the best place to learn React
      </Link>
    </h2>
    <div
      className="list__item__meta"
    >
      <t
        ago={false}
        calendar={false}
        className="post__date"
        date={1467166872634}
        element="time"
        filter={[Function]}
        format="MMMM Do, YYYY"
        fromNow={false}
        interval={60000}
        onChange={[Function]}
        toNow={false}
        unix={false}
        utc={false}
      />
      <span
        className="separator"
      >
         • 
      </span>
      <span>
        thingtwo
      </span>
      <span
        className="separator"
      >
         • 
      </span>
      <span>
        2 comment
        s
      </span>
    </div>
    <Connect(PostScoreContainer)
      id="8xf0y6ziyjabvozdd253nd"
    />
  </div>
  <div
    className="list__item__actions"
  >
    <Link
      className="list__item__actions__action"
      replace={false}
      to="/redux/8xf0y6ziyjabvozdd253nd/edit"
    >
      <FontAwesomeIcon$1
        border={false}
        className=""
        fixedWidth={false}
        flip={null}
        icon={
          Object {
            "icon": Array [
              576,
              512,
              Array [],
              "f044",
              "M402.6 83.2l90.2 90.2c3.8 3.8 3.8 10 0 13.8L274.4 405.6l-92.8 10.3c-12.4 1.4-22.9-9.1-21.5-21.5l10.3-92.8L388.8 83.2c3.8-3.8 10-3.8 13.8 0zm162-22.9l-48.8-48.8c-15.2-15.2-39.9-15.2-55.2 0l-35.4 35.4c-3.8 3.8-3.8 10 0 13.8l90.2 90.2c3.8 3.8 10 3.8 13.8 0l35.4-35.4c15.2-15.3 15.2-40 0-55.2zM384 346.2V448H64V128h229.8c3.2 0 6.2-1.3 8.5-3.5l40-40c7.6-7.6 2.2-20.5-8.5-20.5H48C21.5 64 0 85.5 0 112v352c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48V306.2c0-10.7-12.9-16-20.5-8.5l-40 40c-2.2 2.3-3.5 5.3-3.5 8.5z",
            ],
            "iconName": "edit",
            "prefix": "fas",
          }
        }
        listItem={false}
        mask={null}
        name=""
        pull={null}
        pulse={false}
        rotation={null}
        size={null}
        spin={false}
        symbol={false}
        transform={null}
      />
    </Link>
    <a
      className="list__item__actions__action text-danger pointer"
      onClick={[Function]}
    >
      <FontAwesomeIcon$1
        border={false}
        className=""
        fixedWidth={false}
        flip={null}
        icon={
          Object {
            "icon": Array [
              384,
              512,
              Array [],
              "f00d",
              "M323.1 441l53.9-53.9c9.4-9.4 9.4-24.5 0-33.9L279.8 256l97.2-97.2c9.4-9.4 9.4-24.5 0-33.9L323.1 71c-9.4-9.4-24.5-9.4-33.9 0L192 168.2 94.8 71c-9.4-9.4-24.5-9.4-33.9 0L7 124.9c-9.4 9.4-9.4 24.5 0 33.9l97.2 97.2L7 353.2c-9.4 9.4-9.4 24.5 0 33.9L60.9 441c9.4 9.4 24.5 9.4 33.9 0l97.2-97.2 97.2 97.2c9.3 9.3 24.5 9.3 33.9 0z",
            ],
            "iconName": "times",
            "prefix": "fas",
          }
        }
        listItem={false}
        mask={null}
        name=""
        pull={null}
        pulse={false}
        rotation={null}
        size={null}
        spin={false}
        symbol={false}
        transform={null}
      />
    </a>
  </div>
</li>
`;

exports[`PostListItem renders without crashing 1`] = `
<li
  className="list__item"
>
  <div
    className="post__image"
  >
    <Link
      className="post__icon post__icon--react"
      replace={false}
      to="/react/8xf0y6ziyjabvozdd253nd"
    >
      <img
        alt="react"
        src="react.svg"
      />
    </Link>
  </div>
  <div
    className="list__item__data"
  >
    <h2
      className="post__title"
    >
      <Link
        replace={false}
        to="/react/8xf0y6ziyjabvozdd253nd"
      >
        Udacity is the best place to learn React
      </Link>
    </h2>
    <div
      className="list__item__meta"
    >
      <t
        ago={false}
        calendar={false}
        className="post__date"
        date={1467166872634}
        element="time"
        filter={[Function]}
        format="MMMM Do, YYYY"
        fromNow={false}
        interval={60000}
        onChange={[Function]}
        toNow={false}
        unix={false}
        utc={false}
      />
      <span
        className="separator"
      >
         • 
      </span>
      <span>
        thingtwo
      </span>
      <span
        className="separator"
      >
         • 
      </span>
      <span>
        2 comment
        s
      </span>
    </div>
    <Connect(PostScoreContainer)
      id="8xf0y6ziyjabvozdd253nd"
    />
  </div>
  <div
    className="list__item__actions"
  >
    <Link
      className="list__item__actions__action"
      replace={false}
      to="/react/8xf0y6ziyjabvozdd253nd/edit"
    >
      <FontAwesomeIcon$1
        border={false}
        className=""
        fixedWidth={false}
        flip={null}
        icon={
          Object {
            "icon": Array [
              576,
              512,
              Array [],
              "f044",
              "M402.6 83.2l90.2 90.2c3.8 3.8 3.8 10 0 13.8L274.4 405.6l-92.8 10.3c-12.4 1.4-22.9-9.1-21.5-21.5l10.3-92.8L388.8 83.2c3.8-3.8 10-3.8 13.8 0zm162-22.9l-48.8-48.8c-15.2-15.2-39.9-15.2-55.2 0l-35.4 35.4c-3.8 3.8-3.8 10 0 13.8l90.2 90.2c3.8 3.8 10 3.8 13.8 0l35.4-35.4c15.2-15.3 15.2-40 0-55.2zM384 346.2V448H64V128h229.8c3.2 0 6.2-1.3 8.5-3.5l40-40c7.6-7.6 2.2-20.5-8.5-20.5H48C21.5 64 0 85.5 0 112v352c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48V306.2c0-10.7-12.9-16-20.5-8.5l-40 40c-2.2 2.3-3.5 5.3-3.5 8.5z",
            ],
            "iconName": "edit",
            "prefix": "fas",
          }
        }
        listItem={false}
        mask={null}
        name=""
        pull={null}
        pulse={false}
        rotation={null}
        size={null}
        spin={false}
        symbol={false}
        transform={null}
      />
    </Link>
    <a
      className="list__item__actions__action text-danger pointer"
      onClick={[Function]}
    >
      <FontAwesomeIcon$1
        border={false}
        className=""
        fixedWidth={false}
        flip={null}
        icon={
          Object {
            "icon": Array [
              384,
              512,
              Array [],
              "f00d",
              "M323.1 441l53.9-53.9c9.4-9.4 9.4-24.5 0-33.9L279.8 256l97.2-97.2c9.4-9.4 9.4-24.5 0-33.9L323.1 71c-9.4-9.4-24.5-9.4-33.9 0L192 168.2 94.8 71c-9.4-9.4-24.5-9.4-33.9 0L7 124.9c-9.4 9.4-9.4 24.5 0 33.9l97.2 97.2L7 353.2c-9.4 9.4-9.4 24.5 0 33.9L60.9 441c9.4 9.4 24.5 9.4 33.9 0l97.2-97.2 97.2 97.2c9.3 9.3 24.5 9.3 33.9 0z",
            ],
            "iconName": "times",
            "prefix": "fas",
          }
        }
        listItem={false}
        mask={null}
        name=""
        pull={null}
        pulse={false}
        rotation={null}
        size={null}
        spin={false}
        symbol={false}
        transform={null}
      />
    </a>
  </div>
</li>
`;
